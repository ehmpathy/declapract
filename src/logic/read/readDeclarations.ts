import { UseCaseDeclaration } from '../../domain';
import { UseCasesDeclarationInput } from '../../domain/objects/UseCasesDeclarationInput';
import { readYmlFile } from '../../utils/fileio/readYmlFile';
import { listPathsInDirectory } from '../../utils/filepaths/listPathsInDirectory';
import { UserInputError } from '../UserInputError';
import { readPracticeDeclaration } from './getPracticeDeclaration';

export const readDeclarations = async ({ declarationsDir }: { declarationsDir: string }) => {
  // look at all of the file paths in this directory
  const files = await listPathsInDirectory(declarationsDir);

  // check that there's a `practices` directory
  if (!files.includes('practices'))
    throw new UserInputError('a `practices/` directory must be present in your declarations directory');

  // read practices from each practices directory
  const practiceDirectories = await listPathsInDirectory(`${declarationsDir}/practices`);
  if (!practiceDirectories.length)
    throw new UserInputError('at least one practice needs to be defined in the `practices/` directory');
  const practices = await Promise.all(
    practiceDirectories.map((directory) =>
      readPracticeDeclaration({ declaredPracticeDirectory: `${declarationsDir}/practices/${directory}` }),
    ),
  );

  // check that there's a `use-cases.yml` file
  if (!files.includes('useCases.yml'))
    throw new UserInputError('a `useCases.yml` file must be present in your declarations directory');

  // read and validate the useCases yml file
  const useCasesInput = await (async () => {
    const useCasesYml = await readYmlFile({ filePath: `${declarationsDir}/useCases.yml` });
    try {
      return new UseCasesDeclarationInput(useCasesYml);
    } catch (error) {
      throw new UserInputError('`useCases.yml` file contents failed validation', { potentialSolution: error.message });
    }
  })();

  // now build up the use cases
  const useCases = Object.entries(useCasesInput['use-cases']).map(
    ([name, requestedPractices]) =>
      new UseCaseDeclaration({
        name,
        practices: requestedPractices.map((requestedPracticeKey) => {
          if (!new RegExp('^practice:').test(requestedPracticeKey))
            throw new UserInputError(
              `practice key does not match pattern 'practice:\${practiceName}' for use-case:${name}. key: ${requestedPracticeKey}`,
            );
          const requestedPracticeName = requestedPracticeKey.replace(/^practice:/, '');
          const foundPractice = practices.find((practice) => practice.name === requestedPracticeName);
          if (!foundPractice)
            throw new UserInputError(
              `A use-case specified a practices that was not defined: 'use-case:${name}' specified 'practice:${requestedPracticeName}', but no practice with name '${requestedPracticeName}' has been defined`,
              {
                potentialSolution: 'Have you checked that there is not a typo in the practice name?',
              },
            );
          return foundPractice;
        }),
      }),
  );

  // and return both
  return { useCases, practices };
};
