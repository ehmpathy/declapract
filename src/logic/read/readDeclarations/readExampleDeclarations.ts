import { ExampleDeclaration } from '../../../domain/objects/ExampleDeclaration';
import { listPathsInDirectory } from '../../../utils/filepaths/listPathsInDirectory';

export const readExampleDeclarations = async ({ declaredExamplesDirectory }: { declaredExamplesDirectory: string }) => {
  const exampleDirectories = await listPathsInDirectory(declaredExamplesDirectory);
  const examples = exampleDirectories.map((directory) => {
    const name = directory;
    return new ExampleDeclaration({
      name,
    });
  });
  return examples;
};
