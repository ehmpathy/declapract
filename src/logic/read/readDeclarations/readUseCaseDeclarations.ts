import uniqBy from 'lodash.uniqby';
import { PracticeDeclaration, UseCaseDeclaration } from '../../../domain';
import { ExampleDeclaration } from '../../../domain/objects/ExampleDeclaration';
import { UseCasesDeclarationInput } from '../../../domain/objects/UseCasesDeclarationInput';
import { readYmlFile } from '../../../utils/fileio/readYmlFile';
import { UserInputError } from '../../UserInputError';

export const readUseCaseDeclarations = async ({
  declaredUseCasesPath,
  practices,
}: {
  declaredUseCasesPath: string;
  practices: PracticeDeclaration[];
}) => {
  // try to read the file
  const useCasesYml = await readYmlFile({ filePath: declaredUseCasesPath });

  // read and validate the useCases yml file
  const useCasesInput = await (async () => {
    try {
      return new UseCasesDeclarationInput(useCasesYml);
    } catch (error) {
      throw new UserInputError('use-cases declaration file contents failed validation', {
        potentialSolution: error.message,
      });
    }
  })();

  // now build up the use cases
  const useCases = Object.entries(useCasesInput['use-cases']).map(
    ([name, definition]) =>
      new UseCaseDeclaration({
        name,
        practices: definition.practices.map((requestedPracticeName) => {
          const foundPractice = practices.find((practice) => practice.name === requestedPracticeName);
          if (!foundPractice)
            throw new UserInputError(
              `A use-case specified a practices that was not defined. 'use-case:${name}' specified 'practice:${requestedPracticeName}', but no practice with name '${requestedPracticeName}' has been defined`,
              {
                potentialSolution: 'Have you checked that there is not a typo in the practice name?',
              },
            );
          return foundPractice;
        }),
        example: definition.example ? new ExampleDeclaration({ name: definition.example }) : null,
      }),
  );

  // now hydrate the use cases that requested "extends" of another use case
  const hydratedUseCases = useCases.map((useCase) => {
    // determine whether this use case asked for an extension
    const { extends: extendsUseCaseNames } = useCasesInput['use-cases'][useCase.name];
    if (!extendsUseCaseNames) return useCase; // if it doesn't extend any use cases, then we can stop here

    // for each use case it extends and build up the full, hydrated set of practices
    const hydratedPractices = [...useCase.practices];
    extendsUseCaseNames.forEach((extendsUseCaseName) => {
      // lookup the use case that it extends
      const extendedUseCase = useCases.find((candidateUseCase) => candidateUseCase.name === extendsUseCaseName);
      if (!extendedUseCase)
        throw new UserInputError(
          `A use-case was defined to extend a non-existent use-case. 'use-case:${useCase.name}' was defined to extend 'use-case:${extendsUseCaseName}', but no use case with name '${extendsUseCaseName}' is defined`,
          { potentialSolution: 'Was there a typo?' },
        );

      // push the practices on that extended use case
      hydratedPractices.push(...extendedUseCase.practices);
    });

    // now dedupe them by name
    const uniqueHydratedPractices = uniqBy(hydratedPractices, (practice) => practice.name);

    // and return the hydrated use case
    return new UseCaseDeclaration({ ...useCase, practices: uniqueHydratedPractices });
  });

  // and return the use cases
  return hydratedUseCases;
};
