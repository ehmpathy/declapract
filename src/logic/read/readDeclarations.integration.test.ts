import { readDeclarations } from './readDeclarations';

describe('readDeclarations', () => {
  it('should get the declarations correctly', async () => {
    const declarations = await readDeclarations({
      declarationsDir: `${__dirname}/../__test_assets__/exampleDeclarations`,
    });
    expect(declarations.useCases.length).toEqual(1);
    expect(declarations.practices.length).toEqual(6);
    expect(declarations).toMatchSnapshot();
  });
});
