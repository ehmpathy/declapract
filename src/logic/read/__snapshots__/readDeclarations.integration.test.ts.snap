// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readDeclarations should get the declarations correctly 1`] = `
Object {
  "practices": Array [
    PracticeDeclaration {
      "badPractices": Array [
        CheckProjectDeclaration {
          "checks": Array [
            CheckFileDeclaration {
              "check": [Function],
              "fix": null,
              "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/bad-practices/moment/package.json",
              "required": true,
              "type": "CUSTOM",
            },
          ],
          "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/bad-practices/moment",
          "name": "moment",
          "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
        },
      ],
      "bestPractice": CheckProjectDeclaration {
        "checks": Array [
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/best-practice/package.json",
            "required": true,
            "type": "CUSTOM",
          },
        ],
        "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/best-practice",
        "name": "dates-and-times",
        "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
      },
      "name": "dates-and-times",
    },
    PracticeDeclaration {
      "badPractices": Array [
        CheckProjectDeclaration {
          "checks": Array [
            CheckFileDeclaration {
              "check": [Function],
              "fix": null,
              "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/bad-practices/model-dir/src/model/**/*",
              "required": true,
              "type": "EXISTS",
            },
          ],
          "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/bad-practices/model-dir",
          "name": "model-dir",
          "readme": "we use \`domain\` directory instead of \`model\` directory
- model directory was what we originally went with, but realized \`domain\` was more explicit and better represented what we were going after
",
        },
        CheckProjectDeclaration {
          "checks": Array [
            CheckFileDeclaration {
              "check": [Function],
              "fix": null,
              "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/bad-practices/services-dir/src/services/**/*",
              "required": true,
              "type": "EXISTS",
            },
          ],
          "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/bad-practices/services-dir",
          "name": "services-dir",
          "readme": null,
        },
      ],
      "bestPractice": CheckProjectDeclaration {
        "checks": Array [
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice/src/contract/**/*",
            "required": true,
            "type": "EXISTS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice/src/data/clients/**/*.ts",
            "required": false,
            "type": "CONTAINS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice/src/data/dao/**/*.ts",
            "required": false,
            "type": "EXISTS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice/src/domain/constants",
            "required": false,
            "type": "EXISTS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice/src/domain/index",
            "required": true,
            "type": "EXISTS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice/src/domain/objects/index",
            "required": true,
            "type": "EXISTS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice/src/logic/**/*.ts",
            "required": true,
            "type": "EXISTS",
          },
        ],
        "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/directory-structure-src/best-practice",
        "name": "directory-structure-src",
        "readme": "layers to a project:
- contract layer
  - summary: responsible for strictly exposing the logic/domain defined in the service
  - responsibilities:
    - specifies the inputs and outputs
    - normalizes inputs on behalf of user
    - validates the inputs
    - normalizes outputs on behalf of service
- domain layer
  - summary: defines the domain which this service masters
  - responsibilities:
    - specify domain objects
    - specify constants
- logic layer
  - summary: defines the business logic that this service manages
  - responsibilities:
    - define the business logic of the domain in an organized and maintainable way
- data layer:
  - summary: defines how to persist data / leverage third party clients
  - responsibilities:
    - daos
    - clients
",
      },
      "name": "directory-structure-src",
    },
    PracticeDeclaration {
      "badPractices": Array [],
      "bestPractice": CheckProjectDeclaration {
        "checks": Array [
          CheckFileDeclaration {
            "check": [Function],
            "fix": [Function],
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice/.prettierignore",
            "required": true,
            "type": "EQUALS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice/package.json",
            "required": true,
            "type": "CUSTOM",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": [Function],
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice/prettier.config.js",
            "required": true,
            "type": "EQUALS",
          },
        ],
        "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice",
        "name": "prettier",
        "readme": null,
      },
      "name": "prettier",
    },
    PracticeDeclaration {
      "badPractices": Array [],
      "bestPractice": CheckProjectDeclaration {
        "checks": Array [
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/serverless/best-practice/package.json",
            "required": true,
            "type": "CUSTOM",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/serverless/best-practice/serverless.yml",
            "required": true,
            "type": "CONTAINS",
          },
        ],
        "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/serverless/best-practice",
        "name": "serverless",
        "readme": null,
      },
      "name": "serverless",
    },
    PracticeDeclaration {
      "badPractices": Array [],
      "bestPractice": CheckProjectDeclaration {
        "checks": Array [
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/.gitignore",
            "required": true,
            "type": "CONTAINS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": [Function],
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/.terraform-version",
            "required": true,
            "type": "EQUALS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/dev/main.tf",
            "required": true,
            "type": "CONTAINS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": [Function],
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/dev/versions.tf",
            "required": true,
            "type": "EQUALS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": null,
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/prod/main.tf",
            "required": true,
            "type": "CONTAINS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": [Function],
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/prod/versions.tf",
            "required": true,
            "type": "EQUALS",
          },
          CheckFileDeclaration {
            "check": [Function],
            "fix": [Function],
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/product/variables.tf",
            "required": true,
            "type": "EQUALS",
          },
        ],
        "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice",
        "name": "terraform",
        "readme": "best pattern found for terraform so far is to put different environments into their own directory - and just have them reference a shared product dir
",
      },
      "name": "terraform",
    },
    PracticeDeclaration {
      "badPractices": Array [],
      "bestPractice": CheckProjectDeclaration {
        "checks": Array [
          CheckFileDeclaration {
            "check": [Function],
            "fix": [Function],
            "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/util-sleep/best-practice/src/utils/sleep.ts",
            "required": false,
            "type": "EQUALS",
          },
        ],
        "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/util-sleep/best-practice",
        "name": "util-sleep",
        "readme": null,
      },
      "name": "util-sleep",
    },
  ],
  "useCases": Array [
    UseCaseDeclaration {
      "name": "lambda-service",
      "practices": Array [
        PracticeDeclaration {
          "badPractices": Array [
            CheckProjectDeclaration {
              "checks": Array [
                CheckFileDeclaration {
                  "check": [Function],
                  "fix": null,
                  "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/bad-practices/moment/package.json",
                  "required": true,
                  "type": "CUSTOM",
                },
              ],
              "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/bad-practices/moment",
              "name": "moment",
              "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
            },
          ],
          "bestPractice": CheckProjectDeclaration {
            "checks": Array [
              CheckFileDeclaration {
                "check": [Function],
                "fix": null,
                "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/best-practice/package.json",
                "required": true,
                "type": "CUSTOM",
              },
            ],
            "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/best-practice",
            "name": "dates-and-times",
            "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
          },
          "name": "dates-and-times",
        },
        PracticeDeclaration {
          "badPractices": Array [],
          "bestPractice": CheckProjectDeclaration {
            "checks": Array [
              CheckFileDeclaration {
                "check": [Function],
                "fix": [Function],
                "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice/.prettierignore",
                "required": true,
                "type": "EQUALS",
              },
              CheckFileDeclaration {
                "check": [Function],
                "fix": null,
                "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice/package.json",
                "required": true,
                "type": "CUSTOM",
              },
              CheckFileDeclaration {
                "check": [Function],
                "fix": [Function],
                "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice/prettier.config.js",
                "required": true,
                "type": "EQUALS",
              },
            ],
            "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/prettier/best-practice",
            "name": "prettier",
            "readme": null,
          },
          "name": "prettier",
        },
        PracticeDeclaration {
          "badPractices": Array [],
          "bestPractice": CheckProjectDeclaration {
            "checks": Array [
              CheckFileDeclaration {
                "check": [Function],
                "fix": null,
                "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/serverless/best-practice/package.json",
                "required": true,
                "type": "CUSTOM",
              },
              CheckFileDeclaration {
                "check": [Function],
                "fix": null,
                "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/serverless/best-practice/serverless.yml",
                "required": true,
                "type": "CONTAINS",
              },
            ],
            "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/serverless/best-practice",
            "name": "serverless",
            "readme": null,
          },
          "name": "serverless",
        },
      ],
    },
  ],
}
`;
